### PR TITLE
Extracted `Database` into a separate `fuel-database` crate

### PR DESCRIPTION
**Pull request by xgreenx**
## Overview

It is pure extraction of the `Database` and all related stuff(tables definition, `DatabaseTransaction`, other types) to its own crate. There are no additional changes in the business logic except the small one highlighted in the comment. Also, the change follows the ideology described below and moves the implementation of the DB trait defined by each service into the service.

I accidentally started this change because I needed database primitives to have cross-service database transaction(I meant  `DatabaseTransaction` with some additional traits). 

We already tried to do that in the past during this issue https://github.com/FuelLabs/fuel-core/issues/464. But it was not a good time for this kind of change. With upcoming [architecture refactoring](https://github.com/FuelLabs/fuel-core/issues/809), we need this kind of change if we want finally remove `fuel-core-interfaces`.

In the linked issue, there were some concerns:

> service module crates taking a direct dependency on database / rocksdb, as that will drastically impact our unit testing performance

`rocksdb` is an optional dependency of the `fuel-database` disabled by default. All services will use it only for database primitives/types. Only `fuel-core` will enable it, and only `fuel-core` will decide in runtime which database type to use(In memory/RocksDB/New one). 

All services will use an in-memory database for the unit tests by default. `fuel-database`(in most cases) and integration tests are responsible for testing of how RocksDB works.

> a large amount of churn while there are other critical tasks in flight

Seems it is a good time for this change if we want to remove `fuel-core-interfaces`=) Also, it can remove `MockDB` in each service(and supporting of it). 

> loose design / boundaries, the db-related logic in fuel-core could likely be split into several different crates. More design and justification is necessary here.

`fuel-database` only implements the basic functionality of storage and common business logic(like get block, transactions, receipts), and defines primitives and abstraction. The `fuel-database` as upcoming `fuel-core-types` is a low-level library used by each service. The logic of each service belongs to the service. So each service defines its database trait with all required functionality. The service should implement this trait for `Database` directly or by a generic implementation(by using some traits, for example, `KeyValueStore`). 

The services still want to work with the trait defined by them because it allows them to not depend on `Database` and allows downstream crates to provide their "Database" with their own implementation of traits(at the same time `fuel-core` will use a `Database` as the primary type).

## Follow-ups

This change only moves the code we had before without introducing a new one. But we can make more changes later:

- With a low-level database crate, we can remove the usage of `MockDB` and use the `Database` directly.
- Maybe we want to implement all basic functionality for a generic type that implements `KeyValueStore` instead of `Database`.
- Maybe we want to use a generic database instead of `DataSource` inside services. It will allow having more complex traits with generics(with `dyn` keyword the trait should be [safe](https://doc.rust-lang.org/reference/items/traits.html#object-safety)).
- Change the structure of the `fuel-database` crate because it is the same as it was in the `fuel-core`.
- Review traits and types, maybe change naming, add comments, maybe improve our abstraction.
- Improve our errors during work with the database(because we use several there).

### CODE REVIEW

File: fuel-core/src/service/modules.rs
Comment (xgreenx) [inline fuel-core/src/service/modules.rs:305]: Here I added a new logic(to return a default value)=( Because the `RelayerDB` trait now lives inside of `fuel-relayer`(instead of `fuel-core-interfaces`) I can't use the `get_finalized_da_height` if `relayer` feature is disabled.

File: fuel-database/src/lib.rs
Comment (xgreenx) [inline fuel-database/src/lib.rs:38]: Moving this stuff will affect https://github.com/FuelLabs/fuel-core/pull/796 =( But I will help with it

File: fuel-database/src/error.rs
Comment (ControlCplusControlV) [inline fuel-database/src/error.rs:5]: As a library I believe the correct way would be hierarchical errors, so external facing functions all return a generic `DatabaseError`, which is itself a wrapper for various errors like `IoError` (which may occur writing data) or `InitializationError` (an error wrapper for all initialization related errors) 

Looks like this has been done to some degree, but would be good to finalize an external error enum before merging

File: fuel-database/Cargo.toml
Comment (ControlCplusControlV) [inline fuel-database/Cargo.toml:34]: Default features on `fuel-core` include both `metrics` and `rocksdb`, wouldn't this then be a behavior change?

File: fuel-poa-coordinator/src/db.rs
Comment (ControlCplusControlV) [inline fuel-poa-coordinator/src/db.rs:15]:
```suggestion
    /// Returns error if already sealed
```
Doc comment

File: fuel-relayer/src/db.rs
Comment (ControlCplusControlV) [inline fuel-relayer/src/db.rs:56]: Doesn't a result need to be handled here?

**Review by ControlCplusControlV — CHANGES_REQUESTED** [review]
Some initial changes / questions, sorry if I overlooked something, this is a lot to digest

Also I do understand why we need this change, but at the very least shouldn't this wait until the issues mentioned on Slack (#792 , #710 , #793 , #802 ) be fixed before merging?

Also would be nice to add this to #809 since it has to do with upcoming refactors

File: fuel-database/src/lib.rs
Comment (xgreenx) [reply fuel-database/src/lib.rs:134]: I only moved the code=)

BTW, it is handled in the https://github.com/FuelLabs/fuel-core/pull/643

File: fuel-database/src/lib.rs
Comment (xgreenx) [reply fuel-database/src/lib.rs:129]: Which TODO do you mean?

File: fuel-database/src/lib.rs
Comment (xgreenx) [reply fuel-database/src/lib.rs:153]: I only moved the code, it was here before=)

File: fuel-database/src/lib.rs
Comment (xgreenx) [reply fuel-database/src/lib.rs:137]: What do you mean?

File: fuel-database/src/error.rs
Comment (xgreenx) [reply fuel-database/src/error.rs]: It was before here, but okay)

File: fuel-database/src/error.rs
Comment (xgreenx) [reply fuel-database/src/error.rs:7]: I agree, but this PR is simple movement stuff from one place to another. The issue with errors is mentioned in the `follow-ups` section in the description=)

File: fuel-database/Cargo.toml
Comment (xgreenx) [reply fuel-database/Cargo.toml:34]: No, because `fuel-core` still enables those features by default=)

File: fuel-database/src/tables/resource.rs
Comment (xgreenx) [reply fuel-database/src/tables/resource.rs:217]: It will be done in the follow-up PRs mentioned in the description. I will create them after the merge of this PR and all comments. If we forget about it, it is not critical todo to forget =) It means that no one uses resources.

File: fuel-p2p/Cargo.toml
Comment (xgreenx) [reply fuel-p2p/Cargo.toml:20]: It is a guard for the future to not use unexpected default features. Right now, `fuel-database` hasn't default features.

File: fuel-relayer/src/db.rs
Comment (xgreenx) [reply fuel-relayer/src/db.rs:56]: That code was before=)
image


Of course we need to handle it and return a result, but it is not part of this PR=)

**Review by xgreenx — COMMENTED** [review]
> but at the very least shouldn't this wait until the issues mentioned on Slack 

I tried to fix https://github.com/FuelLabs/fuel-core/issues/791. But seems it is not listed, so yeah, we can wait for a new release first if we don't need it=)

File: fuel-database/src/lib.rs
Comment (ControlCplusControlV) [reply fuel-database/src/lib.rs:129]: There was a TODO I saw but was on an `impl` not trait definition so unrealted, my bad

File: fuel-database/src/error.rs
Comment (ControlCplusControlV) [reply fuel-database/src/error.rs]: Thanks, sorry this just got to be me before and now there is a chance to fix it

File: fuel-database/src/lib.rs
Comment (ControlCplusControlV) [reply fuel-database/src/lib.rs:137]: Since you're moving the code and this isn't a refactor this is fine, but `iter_all` logic is definitely weird right now in the database

File: fuel-database/src/lib.rs
Comment (ControlCplusControlV) [reply fuel-database/src/lib.rs:134]: Ok, as long as its marked down somewhere

**Comment by Voxelot** [comment]
> rocksdb is an optional dependency of the fuel-database disabled by default. All services will use it only for database primitives/types. Only fuel-core will enable it, and only fuel-core will decide in runtime which database type to use(In memory/RocksDB/New one).

My preference would be to avoid taking a dep on rocksdb at all, making the database crate mostly pure / interfaces. This would likely involve the adding the `KeyValueStore` & `BatchOperations` traits as ports to fuel-database.

File: fuel-relayer/src/db.rs
Comment (Voxelot) [reply fuel-relayer/src/db.rs:56]: Didn't you say you were going to fix this already? 😆 

https://github.com/FuelLabs/fuel-core/issues/563#issuecomment-1216085668

File: fuel-database/src/lib.rs
Comment (Voxelot) [reply fuel-database/src/lib.rs:38]: This is an example of why I've been hesitant towards creating a mega-db crate. This shared set of columns seemingly breaks our attempt of having a clean separation of concerns. While implementing mappables and service db ports from the crate perspective is nice, we still end up with a cross-service dependency through these shared columns. Any service that needs to change columns will cascade into potentially breaking adapter changes in all the other crates (even if they had no real logical changes).

While this may not seem like a big issue now when compared to the benefits of sharing the database when considering the executor migration or cross-service, we now open the possibility for implementers of new crates to just start throwing what they need into the shared database crate instead of using their own local ports without really being aware of any difference.

By allowing fuel-core to be responsible for aggregating the database ports from services and having full control over how they are implemented, it resolves the following situations:
 - Allowing custom impls / fine tuning of database traits in the context of a fuel-core node. E.g. using separate instances of rocksdb for merkle tree updates vs plain contract state for performance.
 - Prevent devs from arbitrarily updating a shared database crate whenever they need something custom for their use-case, causing unintended side-affects for other projects using our service crates in different contexts (e.g. putting our block executor into a parachain or cosmos app).

I get how some set of shared data interfaces would be beneficial for things like the executor migration or cross-service txs. An alternative approach to solving these problems without fuel-database or fuel-core-interfaces could be:

- Create an alternative to `DatabaseTransaction` that is entirely trait based. The new trait should have similar AsRef/Deref functionality as before allowing for casts to the db port of each respective service using it.
  - This could be in a new crate as long as it's only highly generic database traits that don't know anything about the underlying services or what their requirements/mappables/columns are. 
- Then services like the executor would just use its own db port trait which defines the mappings it needs, and also use the new trait for database transactions.

This way fuel-core has ultimate control over how db calls get implemented and services don't share any dependency on each other's database requirements.

Comment (freesig) [reply fuel-database/src/lib.rs:38]: I’ve read through this a few times and it’s not really landing. Can you provide a small example of what a trait based transaction could like like?

**Comment by freesig** [comment]
I'm not really convinced we need this change. Are there any other projects that actually currently want to use `fuel-database`?

File: fuel-database/src/lib.rs
Comment (freesig) [reply fuel-database/src/lib.rs:38]: Is the idea to have some kind of custom functions for each request on a service? So instead of using columns and keys we have a getter function for every piece of data we need?

Comment (xgreenx) [reply fuel-database/src/lib.rs:38]: ## Purpose of this PR

I want to highlight one more time that this change only moves stuff to make the review as simple as possible. Real refactoring to be a crate of database primitives will be done in follow-ups. There are two variants of how to create this crate:
- Extract the whole logic into one crate and break it down(to downstream crates) with new changes after.
- Define new primitives in this new crate. After, we will start refactoring existing logic to use those primitives.

I selected the first approach because it allows refactoring with the right primitives that will be easily renamed/replaced later. And it will parallelize refactoring(unblock in some cases because anyone can contribute to the `fuel-database` crate) to use ports/adaptors and new database primitives. 

> This could be in a new crate as long as it's only composed of highly generic database traits that don't know anything about the underlying services or what their requirements/mappables/columns are. 

You already mentioned a new crate where we can store database primitives. It is this crate but designed for `fuel-core` purposes. I will describe "`fuel-core` purposes" below.

## `fuel-core` purposes

The approach with ports and adaptors is good and will provide more flexibility for third parties who want to use our crates.

But at the same time, it means we move most of the logic into `fuel-core`. How to work with the storage is also a business logic of the service. Yes, it is not the responsibility of the service to know about the `RocksDB` or `InMemory` database, but it is the responsibility to update several storage tables, like `Coins` and `Messages`. 

We will face this problem while moving the `executor.rs` into the `fuel-executor` crate. [`ContractRef`](https://github.com/FuelLabs/fuel-core/blob/bad93d1e626aae61cfe9a8ede873f026508cc34b/fuel-block-executor/src/refs/contract.rs#L24) is an example of simplification development and aggregation of the business logic of the service. Moving this into `fuel-core` also transits the business logic of the service there.

My idea is to follow the ports/adapters approach but simultaneously provide a default implementation of the database port for `fuel-core`. Because our development is oriented toward `fuel-core` and support of modularity is an additional feature.

For third parties, it means that they have more ways of customizability:
1. They can reuse our `Database`(`RocksDB` or `InMemory`) and the default implementation of the database port then they have the same logic.
1. They can use their custom "database" but our default implementation of the port.
1. They can implement database port with their own logic.

Without it, they can only `1.` and `3.`.

At the same time, for us, we store the database business logic of the service inside of the service.

It is why it is database primitives for `fuel-core` purposes=)

## Other thoughts after the discussion above

Initially, I wanted to call this crate `fuel-storage`, but we already have the same crate. So I called it the `fuel-database`. But after some thought, it is the wrong naming. The proper name is `fuel-core-storage`.

The `fuel-storage` is a library for database primitives used by `fuel-core` and `fuel-vm`. Because `fuel-vm` is independent(doesn't know about `fuel-core`), it defines its database port. But also defines its `Mappable` types and provides a default implementation for this port.

The same with `fuel-core`. We use primitives from `fuel-storage` and `fuel-vm-storage` to define the required `Mappable` types for the default storage implementation around an unknown `Database`. So `fuel-core-storage` seems a good name, and we can use `fuel-database` for the crate with `Database` implementation(instead of storing it inside of the `fuel-core`).

The `fuel-core-storage` provides storage traits, mappable types, and other traits/types like `Transactional`/`Transaction<Database>`. It will be used by all services.

The `fuel-database` crate will provide an implementation of the `RocksDB`, `InMemory`, `Transactional<InMemory, RocksDB>` databases. It is used only by `fuel-core` in the `[dependencies]` section, and used by all service in the `[dev-dependencies]`

## Answers to questions

> Create an alternative to DatabaseTransaction that is entirely trait based. The new trait should have similar AsRef/Deref functionality as before allowing for casts to the db port of each respective service using it.

> I’ve read through this a few times and it’s not really landing. Can you provide a small example of what a trait based transaction could like like?

We discussed that yesterday on the call=) It is what I plan to implement [here](https://github.com/FuelLabs/fuel-core/compare/master...bugfix/race-condition-seal-block?expand=1) for the `poa-coordinator` service.
 
The idea is: 
Because the `DatabaseTransaction` structure is only known on the `fuel-core` level, we can't use it in the `fuel-block-producer`. But we can use a generic `DbTransaction: Transactional + AsRef<BlockProducerDB> + AsMut<BlockProducerDB>`, that implements transactional functionality(like `commit`) and also return a `Database` instance that implements the database port, defined by the by `fuel-block-producer` crate. 

With generics, it doesn't look lovely like:
```rust
pub trait Transactional {
    fn commit(self) -> anyhow::Result<()>;
}

// It is a database port defined by the `poa-coordiantor`.
pub trait BlockDb: Send + Sync {
    fn block_height(&self) -> Result<BlockHeight>;

    fn seal_block(
        &mut self,
        block_id: BlockId,
        consensus: FuelBlockConsensus,
    ) -> Result<()>;
}

#[async_trait::async_trait]
pub trait BlockProducer<DbTransaction, Database>: Send + Sync
where
    DbTransaction: Transactional + AsRef<Database> + AsMut<Database>,
    Database: BlockDb,
{
    // TODO: Right now production and execution of the block is one step, but in the future,
    //  `produce_block` should only produce a block without affecting the blockchain state.
    async fn produce_and_execute_block(
        &self,
        height: BlockHeight,
        max_gas: Word,
    ) -> anyhow::Result<UncommittedResult<DbTransaction>>;

    async fn dry_run(
        &self,
        transaction: Transaction,
        height: Option<BlockHeight>,
        utxo_validation: Option<bool>,
    ) -> anyhow::Result<Vec<Receipt>>;
}
```

You can see that `BlockProducer<DbTransaction, Database>` will cause the propagation of these two generics inside of `PoACoordinator`.

With `DatabaseTransaction<Database>` inside of `fuel-core-storage`, it would be `BlockProducer<Database>` it looks more friendly. Also, we plan to replace `dyn BlockDB` with `Database` in the future, so it is expected to be generic around `Database`=)

For now, instead of ugly `BlockProducer<DbTransaction, Database>`, we will use dynamic resolution:
```rust
pub trait Transactional {
    fn commit(self) -> anyhow::Result<()>;
}

// It is a database port defined by the `poa-coordiantor`.
pub trait BlockDb: Send + Sync {
    fn block_height(&self) -> Result<BlockHeight>;

    fn seal_block(
        &mut self,
        block_id: BlockId,
        consensus: FuelBlockConsensus,
    ) -> Result<()>;
}

pub trait DatabaseTransaction: Transactional + AsRef<&dyn BlockDb> + AsMut<&mut dyn BlockDb> {}

#[async_trait::async_trait]
pub trait BlockProducer: Send + Sync {
    // TODO: Right now production and execution of the block is one step, but in the future,
    //  `produce_block` should only produce a block without affecting the blockchain state.
    async fn produce_and_execute_block(
        &self,
        height: BlockHeight,
        max_gas: Word,
    ) -> anyhow::Result<UncommittedResult<Arc<dyn DatabaseTransaction>>>;

    async fn dry_run(
        &self,
        transaction: Transaction,
        height: Option<BlockHeight>,
        utxo_validation: Option<bool>,
    ) -> anyhow::Result<Vec<Receipt>>;
}
```

But this approach creates more `Arc` and `dyn` in our code, but doesn't require any ugly/not expected changes=)

Comment (Voxelot) [reply fuel-database/src/lib.rs:38]: > The fuel-core-storage provides storage traits, mappable types, and other traits/types like Transactional/Transaction<Database>. It will be used by all services.
> 
> The fuel-database crate will provide an implementation of the RocksDB, InMemory, Transactional<InMemory, RocksDB> databases. It is used only by fuel-core in the [dependencies] section, and used by all service in the [dev-dependencies]
> 

Splitting into a fuel-core-storage vs fuel-database, where all the heavy deps and implementation logic of the database is separate, mostly resolves my concerns. As fuel-core-storage would just be high-level traits.

> You can see that BlockProducer<DbTransaction, Database> will cause the propagation of these two generics inside of PoACoordinator.

We definitely want to avoid propagating service-specific db traits between services, as it pollutes the domain with unrelated constraints.

**Comment by xgreenx** [comment]
The PR is too stale, so create an issue to implement it later https://github.com/FuelLabs/fuel-core/issues/970

**Pull request closed without merging**
